lib/ecto/adapters/postgres.ex

```diff
@@ -25,6 +25,22 @@ defmodule Ecto.Adapters.Postgres do
 
       YourApp.Repo.all(Queryable, prepare: :unnamed)
 
+  ### Migration options
+
+    * `:migration_lock` - prevent multiple nodes from running migrations at the same
+      time by obtaining a lock. The value `true` will lock migrations by wrapping
+      the entire migration inside a database transaction, including inserting the
+      migration version into the migration source (by default, "schema_migrations").
+      You may alternatively select `:pg_advisory_lock` which has the benefit
+      of allowing concurrent operations such as creating indexes. (default: true)
+
+  When using the `:pg_advisory_lock` migration lock strategy and Ecto cannot obtain
+  the lock due to another instance occupying the lock, Ecto will wait for 5 seconds
+  and then retry up to 100 times. If the retries are exhausted, the migration will
+  fail. This is configurable on the repo with keys
+  `:migration_advisory_lock_retry_interval_ms` and
+  `:migration_advisory_lock_max_tries`.
```

Comment:
`pg_advisory_lock` will block the SQL call, which means is the database connection that could timeout and I found that harder to test; also harder to provide feedback to the user with that this lock is what we're waiting on.

Reply:
Ah I see. So in the previous method the timeout is infinity. I imagine the idea is if there is another transaction performing the migrations you want to make sure it's finished before deploying the app on the current node. So with this change you could enter a state where the current node stops trying the migrations and it also doesn't know whether they have successfully completed. But @josevalim would have better insight into this than me.

Reply:
no you're right, the timeout for migration connections is set to infinity and still is (i confused myself with a test scenario). I'm open to not having a retry feedback loop and letting the migration attempt and loop forever; but I would rather see that in the logs at least and eventually stop trying so that a bad exit code with a clear error message in the logs might alert developers.

I'm thinking of initContainers in kubernetes that run migrations before app boot and it would stop the instance from booting since the database isn't ready. kubernetes would destroy and try again where it would run into the same scenario until the migration is complete. Looping forever might give a false sense of security if the advisory lock is never released for a catastrophic reason.

Reply:
I see what you are saying. It's a bit unintuitive to me to signal problems in this way, because you would never know if it's a problem or a long-running migration. One of the main advantages of advisory locks is safely being able to perform `concurrently` migrations, which have much less predictable durations.

I would personally use `timeout: :infinity` and then monitor for long-running transactions and # of nodes running my application separately.  If I had a migration in which I don't want it to block my app from starting I would set `lock_timeout` to something very small.

Reply:
I am on the fence on which approach to take but given just rolling with `timeout: infinity` is actually less code, let's go down with that approach. Also LOCK TABLE can lock for similar reasons and we don't take special precaution here, so using `timeout: :infinity` will be more consistent too. :)

Reply:
Sorry for necroing this after merging. I was trying to wrap my head around this and think I see the issue. It doesn't matter but just in case it'srevisited:

```elixir
defp advisory_lock(meta, opts, lock, fun) do
    blocking_queries(meta, lock)
    Ecto.Adapters.SQL.query(meta, "SELECT pg_advisory_lock(#{lock})", [], opts)
    try do
      fun.()
    after
      {:ok, _} = Ecto.Adapters.SQL.query(meta, "SELECT pg_advisory_unlock(#{lock})", [], opts)
      blocking_queries(meta, lock)
    end
end
```

I'm pretty sure the issue is that the get advisory lock -> perform migration -> release advisory lock is not within one connection (the default pool is 10 for the tests so it can use a different connection per statement).  It shouldn't be possible for multiple nodes to perform the same migration because `fun.()` will recalculate the versions. And `fun.()` should never be called until the advisory lock is unblocked, which should only happen for one node at a time. It should be no different than opening a transaction and getting a lock on the `schema_migration` table.

Reply:
I don't think `fun.()` is reading the schema version in this mentioned block according to the logged lines.

The entrypoint of migrations is here in `Ecto.Migrator.run`. In this function it reads the current schema migration versions. Notice it takes a migration lock here.
https://github.com/elixir-ecto/ecto_sql/blob/086968634e1b59d65fbecd59cc6c3385bc7b368f/lib/ecto/migrator.ex#L421-L435

It then passes that downwards until it attempts to apply each of those migrations. 

When attempting to apply the migration, it obtains another migration lock:
https://github.com/elixir-ecto/ecto_sql/blob/086968634e1b59d65fbecd59cc6c3385bc7b368f/lib/ecto/migrator.ex#L679-L683

At this point, another competing node might have already locked and applied the migration.

looking at the log output when using the blocking pg_advisory_lock call, this indeed seems to be happening.

```
11:50:53.013 [debug] QUERY OK db=1.0ms
SELECT pg_advisory_lock(112878997) []

11:50:53.033 [debug] QUERY OK db=1.5ms
SELECT pg_advisory_unlock(112878997) []

11:50:53.058 [debug] QUERY OK db=0.2ms
SELECT pg_advisory_lock(112878997) []

11:50:53.075 [info]  == Running 20220720153111 MigrateTest.Repo.Migrations.LengthyIndex.change/0 forward

11:50:53.078 [info]  create index big_table_foo_index

11:50:58.209 [debug] QUERY OK db=5129.4ms queue=0.6ms idle=44.0ms
CREATE INDEX CONCURRENTLY "big_table_foo_index" ON "big_table" ("foo") []

11:50:58.213 [info]  == Migrated 20220720153111 in 5.1s

11:50:58.298 [debug] QUERY OK db=16.1ms queue=0.5ms idle=54.2ms
INSERT INTO "schema_migrations" ("version","inserted_at") VALUES ($1,$2) [20220720153111, ~N[2022-07-22 15:50:58]]
↳ anonymous fn/6 in Ecto.Migrator.async_migrate_maybe_in_transaction/7, at: lib/ecto/migrator.ex:320

11:50:58.299 [debug] QUERY OK db=0.3ms
SELECT pg_advisory_unlock(112878997) []
```

Here's an example of when I have multiple migrations to be ran. Notice that the first lock/unlock is reading the schema table and doesn't re-check when running later migrations (only locks for applying the migration).

```
11:55:21.002 [debug] QUERY OK db=0.2ms
SELECT pg_advisory_lock(112878997) []

11:55:21.043 [debug] QUERY OK db=0.3ms
SELECT pg_advisory_unlock(112878997) []

11:55:21.111 [debug] QUERY OK db=0.2ms
SELECT pg_advisory_lock(112878997) []

11:55:21.127 [debug] QUERY OK db=0.1ms idle=79.1ms
begin []
↳ Ecto.Migrator.run_maybe_in_transaction/5, at: lib/ecto/migrator.ex:337

11:55:21.137 [info]  == Running 20220720153104 MigrateTest.Repo.Migrations.BigTable.up/0 forward

11:55:21.137 [info]  create table big_table

11:55:21.157 [debug] QUERY OK db=18.8ms
CREATE TABLE "big_table" ("id" bigserial, "foo" text, PRIMARY KEY ("id")) []

11:55:21.161 [info]  execute "INSERT INTO big_table(foo)\n  SELECT md5(random()::text)\n  FROM GENERATE_SERIES(1,1000000) i\n"

11:55:27.064 [debug] QUERY OK db=5.9e3ms
INSERT INTO big_table(foo)
  SELECT md5(random()::text)
  FROM GENERATE_SERIES(1,1000000) i
 []

11:55:27.070 [info]  == Migrated 20220720153104 in 5.9s

11:55:27.151 [debug] QUERY OK db=3.2ms
INSERT INTO "schema_migrations" ("version","inserted_at") VALUES ($1,$2) [20220720153104, ~N[2022-07-22 15:55:27]]
↳ anonymous fn/6 in Ecto.Migrator.async_migrate_maybe_in_transaction/7, at: lib/ecto/migrator.ex:320

11:55:27.189 [debug] QUERY OK db=37.8ms
commit []
↳ Ecto.Migrator.run_maybe_in_transaction/5, at: lib/ecto/migrator.ex:337

11:55:27.192 [debug] QUERY OK db=0.5ms
SELECT pg_advisory_unlock(112878997) []

11:55:27.193 [debug] QUERY OK db=0.2ms
SELECT pg_advisory_lock(112878997) []

11:55:27.195 [info]  == Running 20220720153111 MigrateTest.Repo.Migrations.LengthyIndex.change/0 forward

11:55:27.196 [info]  create index big_table_foo_index

11:55:32.228 [debug] QUERY OK db=5031.0ms queue=0.8ms idle=4.8ms
CREATE INDEX CONCURRENTLY "big_table_foo_index" ON "big_table" ("foo") []

11:55:32.229 [info]  == Migrated 20220720153111 in 5.0s

11:55:32.240 [debug] QUERY OK db=9.8ms queue=0.6ms idle=0.5ms
INSERT INTO "schema_migrations" ("version","inserted_at") VALUES ($1,$2) [20220720153111, ~N[2022-07-22 15:55:32]]
↳ anonymous fn/6 in Ecto.Migrator.async_migrate_maybe_in_transaction/7, at: lib/ecto/migrator.ex:320

11:55:32.241 [debug] QUERY OK db=0.3ms
SELECT pg_advisory_unlock(112878997) []
``` 

I think using a blocking pg_advisory_lock call would work if the entire migration (schema version check and applying the migration) was in a single connection and advisory lock, but that will require some bigger reworking of the Ecto migrator. Another less efficient way is to re-read the schema version before applying the migration as well.

But at this point I'm not sure the trade-off is worth it? Both solutions would work I think, but the non-blocking call gives us an opportunity to provide feedback in logs, not hog a connection while waiting on the lock, and also attach a retry ceiling for how long to wait on the advisory lock.

Reply:
Oh yes sorry I don't think it's worth changing at this point. I was party trying to clear the fog from my head and partly trying to put some info for posterity, if this topic is ever resurfaced.

You can see here that the function passed to the adapter's `lock_for_migrations/3` will recalculate the versions https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/migrator.ex#L533

```elixir
migration_repo.__adapter__().lock_for_migrations(meta, opts, fn ->
  fun.(config, migration_repo.all(query, all_opts))
end)
```

If it didn't work this way then the previous method would also have the same issue.